notes/books/designing-data-intensive-apps.md

```diff
@@ -0,0 +1,122 @@
+as a religion for miraculous promises of eternal scaling.
+
+This also reminds me a lot of the dissertation [Architectural Styles and
+the Design of Network-based Software Architectures](https://www.ics.uci.edu/~fielding/pubs/dissertation/top.htm)
+where REST is defined and a great emphasis is given on not focusing on a single
+architectural style as an answer to all problems, because there is no such thing.
+It also provides some really cool analysis of different architectural styles,
+and a whole framework on how to think about them and compare them. Not to
+find a winner, but to understand which one is applicable to which kind of
+problems. And as the whole Twitter system teaches, you probably won't be able
+to build something scalable with a single architectural style, you need to
+know a lot of them and compose them together.
+
+All this also reminds me of the design principle of
+[Form Follows Function](https://en.wikipedia.org/wiki/Form_follows_function).
+
+## Data Models
+
+The book is also brilliant on how it presents different data models,
+like comparing SQL/NoSQL, because it also avoids finding "winners" and it
+also provides a lot of historical context, specially in detailing that document
+databases are not a new thing, they are actually older than relational databases
+and they just made a comeback because they are much easier to scale,since the
+data model is simpler.
+
+Well, I mean, document databases also make some things easier, like if you always
+retrieved data aggregated in the exact same way a document database will be faster
+and will feel much simpler, because it is. Document databases explore locality
+better. If you always need the very same tree of data, you are golden with
+documents. The hierarchical model represents well one to many relationships,
+like all information related to a user in the same document. The problem is when
+you start having many to many or many to one relationships.
+
+An example is addresses, more than one people can leave on the exact same
+apartment, how to model that many to one relationship in a way that avoids
+duplication and is consistent and easy to maintain ? 
+
+Here enters the relational model, exactly to model that. You could argue that
+we need better models than the relational model, but an hierarchical model
+is not new neither more expressive, it is a simpler model suited for simpler
+problems, and it was well know by the people who developed an alternative
+solution, to when you problem domain demands something more expressive.
+
+I also find remarkable how the whole idea of the relational model is to
+enforce decoupling. They wanted to decoupled programs not just form the internal
+representation of the data, but also from external presentation, so even published
+schemas should be able to change and be extended without requiring changes
+on programs, and yet a lot of people present SQL databases as some sort of
```

Comment:
> a lot of people present SQL databases as some sort of nightmare to evolve

They do have a point, though. I've been spending some time writing migrations with [Alembic](https://alembic.sqlalchemy.org/en/latest/) at work recently, and migration tools like that are really not a good solution to the problem. Martin actually makes an interesting point later in the book (chapter 4) about how you really need forward and backward compatibility to be able to evolve a data format -- not just for network protocols, but also for databases. The way database migrations are usually done (with migration scripts) gives you neither :cry: 

But you're right that we shouldn't blame the relation model for that.

Reply:
yeah I dont have good solutions to this problem either...what I sensed with the whole NoSQL thing is that people make the problem look easy (and blaming schemas as a source of complexity) by completely ignoring the problem and hoping for the best =P. Like creating messy data schemas with the "schemaless" philosophy and code that has to circumvent that and implement the backward/forward compatibility logic anyway, but without a proper/consistent schema. It does feel a lot like dynamic typing the schemaless thing, it makes some things really easier at first but as soon as things get though you are very screwed at runtime =P.

I remember reading about migration strategies on document databases that involved incrementally migrating data as you find data that has an outdated schema. For a single instance of this feels kinda simple, but on the long run it feels even worse than all the SQL migrations thing, because at least with migrations you have some explicit code to migrate data schemas, the document/incremental approach the code that does that can be spread anywhere in the system since it is done dynamically/on the fly and you are never sure when you can actually remove it since the migration/evolution is constant and can take a long time.

Overall I would be prone to backward/forward compatibility, but doing that without duplicating data seems challenging...or maybe if you have a good data model evolving things can be possible...I feel quite incompetent with this because it is a very hard topic and I suspect much more people are equally incompetent =P.

Reply:
and now I'm looking forward for chapter 4 :100: